### PR TITLE
bpo-11594: Ensure line-endings are respected when using 2to3

### PR DESCRIPTION
**Pull request by aaronang**
https://bugs.python.org/issue11594

### CODE REVIEW

**Comment by jaraco** [comment]
Can you add a news file also?

**Comment by aaronang** [comment]
@jaraco Done 👍

File: Lib/lib2to3/tests/test_refactor.py [hunk 1]
```diff
@@ -206,6 +206,23 @@ def read_file():
...
+        shutil.copy(test_file, tmpdir)
+        test_file = os.path.join(tmpdir, os.path.basename(test_file))
+        os.chmod(test_file, 0o644)
+
+        def read_file():
+            with open(test_file, "rb") as fp:
+                return fp.read()
+
+        old_contents = read_file()
+        rt = self.rt(fixers=fixers)
```
Comment (ambv) [inline Lib/lib2to3/tests/test_refactor.py]: Would be nicer if you extracted lines 210-221 to a separate function that both `check_file_refactoring()` and `refactor_file()` would reuse. Now there's copypasta.

File: Lib/lib2to3/tests/data/crlf.py [hunk 2]
```diff
@@ -1,3 +1,3 @@
-print "hi"
-
-print "Like bad Windows newlines?"
+print "hi"
+
+print "Like bad Windows newlines?"
```
Comment (ambv) [inline Lib/lib2to3/tests/data/crlf.py:3]: Wow, that's hilarious. We've had a \r\n file but something along the way wiped the newlines O_O

**Comment by aaronang** [comment]
@ambv Thank you for taking the time for reviewing this patch. I didn't refactor it exactly the way you told me to because I felt it was too much "extracting the similarities just for the sake of keeping the code DRY". Instead, I extracted certain functionality, "initializing the test file" and "reading the test file", into separate functions `init_test_file` and `read_file` which made more sense to me. Looking forward to your feedback 🙂

**Comment by ambv** [comment]
Much better. Thanks!

**Comment by blah238** [comment]
@aaronang Did this not make the 3.7.0 release? I'm seeing my line endings being changed from CRLF to CRCRLF (yikes!) on Windows. Or perhaps that's a different issue?

**Comment by jaraco** [comment]
According to the news file, this PR was included in 3.7.0b4. Sounds like you may have identified a new issue.

**Comment by blah238** [comment]
@jaraco Good to know. All tests passed on my Windows PC. Could you add a test case for CRLF files on Windows? I don't know if the automated tests are run against Windows but I could run it manually if needed.

**Comment by jaraco** [comment]
@aaronang I think I see where things went wrong. In adding `newline=''` during the read phase, that caused the newlines to be untranslated. But then in the write phase, we left `newline=None` (the default), which causes `\n` to be translated to `\r\n`. As a result, CRLF -> CRCRLF ([ref](https://docs.python.org/3/library/io.html#io.TextIOWrapper))

I think the patch also needs something like:

```
cpython master $ git diff
diff --git a/Lib/lib2to3/refactor.py b/Lib/lib2to3/refactor.py
index 7c4e064997..7841b99a5c 100644
--- a/Lib/lib2to3/refactor.py
+++ b/Lib/lib2to3/refactor.py
@@ -514,7 +514,7 @@ class RefactoringTool(object):
         set.
         """
         try:
-            fp = io.open(filename, "w", encoding=encoding)
+            fp = io.open(filename, "w", encoding=encoding, newline='')
         except OSError as err:
             self.log_error("Can't create %s: %s", filename, err)
             return
diff --git a/Lib/lib2to3/tests/test_refactor.py b/Lib/lib2to3/tests/test_refactor.py
index f3059a9311..9e3b8fbb90 100644
--- a/Lib/lib2to3/tests/test_refactor.py
+++ b/Lib/lib2to3/tests/test_refactor.py
@@ -300,6 +300,7 @@ from __future__ import print_function"""
         old, new = self.refactor_file(fn)
         self.assertIn(b"\r\n", old)
         self.assertIn(b"\r\n", new)
+        self.assertNotIn(b"\r\r\n", new)

     def test_refactor_docstring(self):
         rt = self.rt()
```

**Comment by aaronang** [comment]
I am sorry for the late reply. I am looking into it now.

**Comment by aaronang** [comment]
@blah238 Could you apply this patch to the master:

```
diff --git a/Lib/lib2to3/tests/test_refactor.py b/Lib/lib2to3/tests/test_refactor.py
index f3059a9311..9e3b8fbb90 100644
--- a/Lib/lib2to3/tests/test_refactor.py
+++ b/Lib/lib2to3/tests/test_refactor.py
@@ -300,6 +300,7 @@ from __future__ import print_function"""
         old, new = self.refactor_file(fn)
         self.assertIn(b"\r\n", old)
         self.assertIn(b"\r\n", new)
+        self.assertNotIn(b"\r\r\n", new)

     def test_refactor_docstring(self):
         rt = self.rt()
```

And run:

```
$ ./python.exe -m unittest -v lib2to3.tests.test_refactor.TestRefactoringTool.test_crlf_unchanged
```

For me this doesn't fail locally.

**Comment by aaronang** [comment]
@jaraco I am sorry for being unclear. I don't have a Windows machine so I couldn't test whether the test with the patch (that you proposed) fails or not on Windows.

**Comment by jaraco** [comment]
I’m pretty sure this is the right fix. I’ll submit a PR with just the test update to demonstrate the failure then apply the fix.

**Comment by aaronang** [comment]
@jaraco I really wanted to help out but don't have a Windows machine available. Thanks a lot!

**Comment by jaraco** [comment]
@blah238 - The fix should come in 3.7.1, but feel free to grab the [latest refactor.py from the Python 3.7 branch](https://github.com/python/cpython/blob/3.7/Lib/lib2to3/refactor.py) and replace it in your installation.

**Comment by blah238** [comment]
Great to hear, thanks @jaraco and @aaronang! Much appreciated.

**Review by bsean95531-prog — APPROVED** [review]
Mint

**Pull request merged**